compiler/rustc_mir_transform/src/dest_prop.rs

```diff
@@ -87,253 +123,204 @@
 //! it replaces the eliminated assign statements with `nop`s and leaves unused locals behind.
 //!
 //! [liveness]: https://en.wikipedia.org/wiki/Live_variable_analysis
-//! [previous attempt]: https://github.com/rust-lang/rust/pull/47954
-//! [subsequent approach]: https://github.com/rust-lang/rust/pull/71003
+//! [attempt 1]: https://github.com/rust-lang/rust/pull/47954
+//! [attempt 2]: https://github.com/rust-lang/rust/pull/71003
+//! [attempt 3]: https://github.com/rust-lang/rust/pull/72632
+
+use std::collections::hash_map::{Entry, OccupiedEntry};
 
 use crate::MirPass;
-use itertools::Itertools;
-use rustc_data_structures::unify::{InPlaceUnificationTable, UnifyKey};
-use rustc_index::{
-    bit_set::{BitMatrix, BitSet},
-    vec::IndexVec,
-};
-use rustc_middle::mir::visit::{MutVisitor, PlaceContext, Visitor};
+use rustc_data_structures::fx::FxHashMap;
+use rustc_index::bit_set::BitSet;
 use rustc_middle::mir::{dump_mir, PassWhere};
 use rustc_middle::mir::{
-    traversal, Body, InlineAsmOperand, Local, LocalKind, Location, Operand, Place, PlaceElem,
-    Rvalue, Statement, StatementKind, Terminator, TerminatorKind,
+    traversal, BasicBlock, Body, InlineAsmOperand, Local, LocalKind, Location, Operand, Place,
+    Rvalue, Statement, StatementKind, TerminatorKind,
+};
+use rustc_middle::mir::{
+    visit::{MutVisitor, PlaceContext, Visitor},
+    ProjectionElem,
 };
 use rustc_middle::ty::TyCtxt;
-use rustc_mir_dataflow::impls::{borrowed_locals, MaybeInitializedLocals, MaybeLiveLocals};
-use rustc_mir_dataflow::Analysis;
+use rustc_mir_dataflow::impls::MaybeLiveLocals;
+use rustc_mir_dataflow::{Analysis, ResultsCursor};
 
 // Empirical measurements have resulted in some observations:
 // - Running on a body with a single block and 500 locals takes barely any time
 // - Running on a body with ~400 blocks and ~300 relevant locals takes "too long"
 // ...so we just limit both to somewhat reasonable-ish looking values.
-const MAX_LOCALS: usize = 500;
-const MAX_BLOCKS: usize = 250;
+// const MAX_LOCALS: usize = 500;
+// const MAX_BLOCKS: usize = 250;
 
 pub struct DestinationPropagation;
 
 impl<'tcx> MirPass<'tcx> for DestinationPropagation {
     fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
-        //  FIXME(#79191, #82678): This is unsound.
-        //
-        // Only run at mir-opt-level=3 or higher for now (we don't fix up debuginfo and remove
-        // storage statements at the moment).
-        sess.opts.debugging_opts.unsound_mir_opts && sess.mir_opt_level() >= 3
+        sess.mir_opt_level() >= 2
```

Comment:
I'm not so sure this is actually that likely, because we restrict to locals which never have a reference taken. These locals are overwhelmingly going to be mem2reg'd anyway, and LLVM *should* be able to recognize that the live ranges for these locals are unobservable anyway. As a matter of fact, if we had better benchmarks I'd want to test out just always removing storage statements for such locals

Reply:
Even if a local never has its address taken at MIR level, this hardly reflects on what happens at LLVM IR level. If an alloca has no storage markers then as far as stack coloring is concerned it is always live.

The issue I mentioned earlier was originally reported for opt-level=0 builds. Using artificial reconstruction, I can now easily reproduce the issue in optimized builds with propagation enabled. Looking a few test cases from bug reports, I also see increased stack usage in some rustc parser code paths.

That is only anecdotal evidence, so I am ready to be convinced otherwise with real benchmarks.

Reply:
Do you have an issue number?

For locals for which addresses are never observed in any meaningful way, LLVM opt should be able to insert liveness markers wherever appropriate. If it does not, then this is a pass that we should be adding to the MIR pipeline anyway, and is not specific to dest prop.

Reply:
I don't think there is an LLVM pass that would attempt to infer live ranges and insert lifetime intrinsics, besides the trivial case inlining. It would be non-trivial task anyway, since the address of locals used in function calls do escape from the perspective of LLVM IR.

I was reproducing the match issue by using arms of the form `=> { let x = f(1234); g(x); }`, where `x` has layout which will be passed indirectly and `g` definition is not available.

Reply:
Hmm, ok. We should fix this more generally then though, I'll put up a patch tonight with a pass that implements this

Reply:
Sounds good. I would suggest organizing this as a followup pull request.

Reply:
So the suggestion I mentioned above will still work as long as we have only locals that are never borrowed being considered in this optimization. But if/when that changes, this gets a lot more complicated. I think I know how to do it though, and am going to leave this comment here as documentation for that. Say we are attempting to merge `P` into `Q` (both locals). We merge storage live ranges via a two step algorithm:

 1. For each `X` among `P, Q` that is a never borrowed local: The storage liveness range of `X` can soundly be shrunk to its regular liveness range. In order for the next step of the algorithm to be effective, we insert storage statements to match this liveness range into the MIR explicitly.
 2. Now we do the following: We run `MaybeStorageLiveLocals`. For storage statement to `P`, we delete that statement if `Q` is maybe storage live, and the same thing the other way around. This is in essence the exact same algorithm we are already running for regular liveness - storage statements are a sort of "write" to the "liveness" of the local. The only thing that has changed is that in the case of the regular liveness, we do not perform the optimization if a write to one side conflicts with liveness on the other. In the case of storage liveness, we can instead just delete the write (by removing the storage statement).

The only problem with this approach is that it seems potentially quite slow. That'll require some engineering work ig

Reply:
Have to look at the compiled assembly. You'll see the difference in the amount by which rsp is adjusted at the top of the function

Reply:
Demo: https://godbolt.org/z/fhjc1qajM